### DISCUSSION TITLE
How to see what scopes user approved?

### DISCUSSION THREAD

**Discussion by the10xProcrastinator** [nextauthjs/next-auth#6748] [Help] [answered]
### Question 💬

When going through the sign in flow, I have specified a couple scopes for a particular provider. When the user is redirected to the provider to sign in, the user has the option to keep or uncheck some of the scopes. My question is if there's a way to figure out what scopes the user ended up approving? I logged out the data from the redirect() and session() call back but did not see anything to do with scopes there.

### How to reproduce ☕️

```
providers: [
     GenericProviderHere({
          authorization: {
                scope:"profile:read_all, chats:read_all, etc..."
           }
     })
]
```

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by cloudnothings** [comment, marked as answer]
It sounds like you're attempting to use the user's access token from your given provider correct?
This leads me to believe you're using the JWT callback. 
The JWT callback's object can be destructured into { token: {}, user: {}, account: {}, profile: {}, and isNewUser: boolean

When your user first logs in, the JWT callback's {account} property should contain the scope. 

Here's an example of my entire jwt object, sanitized of course:
```
{
  token: {
    name: 'clouds',
    email: 'carlos@rangel.us',
    picture: undefined,
    sub: 'cloud'
  },
  user: {
    id: 'cloud',
    name: 'cloud',
    email: 'carlos@rangel.us',
    image: undefined
  },
  account: {
    provider: 'spotify',
    type: 'oauth',
    providerAccountId: 'cloudn0things',
    access_token: 'jksdhjskdh',
    token_type: 'Bearer',
    expires_at: 1676575195,
    refresh_token: 'jdskldjlskI',
    scope: 'playlist-read-private playlist-read-collaborative user-library-read user-read-currently-playing user-read-email user-read-recently-played user-top-read'      
  },
  profile: {
    display_name: 'cloud',
    email: 'carlos@rangel.us',
    external_urls: { spotify: 'https://open.spotify.com/user/cloud' },       
    followers: { href: null, total: 2 },
    href: 'https://api.spotify.com/v1/users/cloud',
    id: 'cloud',
    images: [],
    type: 'user',
    uri: 'spotify:user:cloud'
  },
  isNewUser: undefined
}
```

You see that my account object has a field called scope filled out. 

The JWT callback only provides this on the initial logins usually, so you can do _include_ something like this to track the scope:

```
callbacks: {
    jwt({ token, account, profile }) {
      if (account) {
        return {
          scope: account.scope
        // other properties you care about
        };
      }
    }
}
```

**Reply by the10xProcrastinator** [reply]
Hey Carlos, thanks for the reply! Sorry I should have been more clear, I'm actually using DB sessions. I need to get the scopes, because the access_token that gets stored in the database is used later on to retrieve data on the user. The goal here is to make sure that the user granted access to the (optional) requested scopes, so that downstream the application doesn't try to fetch data from an API with an access_token that wasn't given the requested permissions. Any advice here for this situation?
